diff --git a/packer/readme.md b/packer/readme.md
new file mode 100644
index 0000000..d908932
--- /dev/null
+++ b/packer/readme.md
@@ -0,0 +1 @@
+copy over the roles folder into packer/shreyach13, then create an ansible.cfg file in packer folder and write :  [defaults] /n roles_path = ./roles. add extra arguments to make it sudo. Then it will run.
diff --git a/terraform/shreyach13/main.tf b/terraform/shreyach13/main.tf
new file mode 100644
index 0000000..a6a2bb5
--- /dev/null
+++ b/terraform/shreyach13/main.tf
@@ -0,0 +1,13 @@
+provider "aws" {
+  region = "eu-central-1"
+}
+
+#This is create a new bucket
+resource "aws_s3_bucket" "shreyatfstate" {
+ bucket = "shreyaspace-tfstore"
+ region = "eu-central-1"
+ tags = {
+    Name  = "Shreya bucket"
+    owner = var.owner
+  }
+}
diff --git a/terraform/shreyach13/production.tfvars b/terraform/shreyach13/production.tfvars
new file mode 100644
index 0000000..e07c5c5
--- /dev/null
+++ b/terraform/shreyach13/production.tfvars
@@ -0,0 +1 @@
+owner = "I am overwrite variable.tf Owner"
diff --git a/terraform/shreyach13/terraform.tfstate b/terraform/shreyach13/terraform.tfstate
new file mode 100644
index 0000000..363bd21
--- /dev/null
+++ b/terraform/shreyach13/terraform.tfstate
@@ -0,0 +1,331 @@
+{
+  "version": 4,
+  "terraform_version": "0.12.9",
+  "serial": 17,
+  "lineage": "37933da2-09f3-c391-ce85-b57d0901dacc",
+  "outputs": {
+    "print_shreya_webserver": {
+      "value": {
+        "ami": "ami-00aa4671cbf840d82",
+        "arn": "arn:aws:ec2:eu-central-1:426714351745:instance/i-02819c8496758a607",
+        "associate_public_ip_address": true,
+        "availability_zone": "eu-central-1b",
+        "cpu_core_count": 1,
+        "cpu_threads_per_core": 1,
+        "credit_specification": [
+          {
+            "cpu_credits": "standard"
+          }
+        ],
+        "disable_api_termination": false,
+        "ebs_block_device": [],
+        "ebs_optimized": false,
+        "ephemeral_block_device": [],
+        "get_password_data": false,
+        "host_id": null,
+        "iam_instance_profile": "",
+        "id": "i-02819c8496758a607",
+        "instance_initiated_shutdown_behavior": null,
+        "instance_state": "running",
+        "instance_type": "t2.micro",
+        "ipv6_address_count": 0,
+        "ipv6_addresses": [],
+        "key_name": "",
+        "monitoring": false,
+        "network_interface": [],
+        "network_interface_id": null,
+        "password_data": "",
+        "placement_group": "",
+        "primary_network_interface_id": "eni-049aeb7aea7b8e2af",
+        "private_dns": "ip-172-31-42-116.eu-central-1.compute.internal",
+        "private_ip": "172.31.42.116",
+        "public_dns": "ec2-18-185-127-109.eu-central-1.compute.amazonaws.com",
+        "public_ip": "18.185.127.109",
+        "root_block_device": [
+          {
+            "delete_on_termination": true,
+            "encrypted": false,
+            "iops": 100,
+            "kms_key_id": "",
+            "volume_id": "vol-072498c51d755da1c",
+            "volume_size": 8,
+            "volume_type": "gp2"
+          }
+        ],
+        "security_groups": [
+          "default"
+        ],
+        "source_dest_check": true,
+        "subnet_id": "subnet-03e6e87e",
+        "tags": {
+          "Name": "ShreyaSpace",
+          "owner": "shreya"
+        },
+        "tenancy": "default",
+        "timeouts": null,
+        "user_data": null,
+        "user_data_base64": null,
+        "volume_tags": {},
+        "vpc_security_group_ids": [
+          "sg-395d185a"
+        ]
+      },
+      "type": [
+        "object",
+        {
+          "ami": "string",
+          "arn": "string",
+          "associate_public_ip_address": "bool",
+          "availability_zone": "string",
+          "cpu_core_count": "number",
+          "cpu_threads_per_core": "number",
+          "credit_specification": [
+            "list",
+            [
+              "object",
+              {
+                "cpu_credits": "string"
+              }
+            ]
+          ],
+          "disable_api_termination": "bool",
+          "ebs_block_device": [
+            "set",
+            [
+              "object",
+              {
+                "delete_on_termination": "bool",
+                "device_name": "string",
+                "encrypted": "bool",
+                "iops": "number",
+                "kms_key_id": "string",
+                "snapshot_id": "string",
+                "volume_id": "string",
+                "volume_size": "number",
+                "volume_type": "string"
+              }
+            ]
+          ],
+          "ebs_optimized": "bool",
+          "ephemeral_block_device": [
+            "set",
+            [
+              "object",
+              {
+                "device_name": "string",
+                "no_device": "bool",
+                "virtual_name": "string"
+              }
+            ]
+          ],
+          "get_password_data": "bool",
+          "host_id": "string",
+          "iam_instance_profile": "string",
+          "id": "string",
+          "instance_initiated_shutdown_behavior": "string",
+          "instance_state": "string",
+          "instance_type": "string",
+          "ipv6_address_count": "number",
+          "ipv6_addresses": [
+            "list",
+            "string"
+          ],
+          "key_name": "string",
+          "monitoring": "bool",
+          "network_interface": [
+            "set",
+            [
+              "object",
+              {
+                "delete_on_termination": "bool",
+                "device_index": "number",
+                "network_interface_id": "string"
+              }
+            ]
+          ],
+          "network_interface_id": "string",
+          "password_data": "string",
+          "placement_group": "string",
+          "primary_network_interface_id": "string",
+          "private_dns": "string",
+          "private_ip": "string",
+          "public_dns": "string",
+          "public_ip": "string",
+          "root_block_device": [
+            "list",
+            [
+              "object",
+              {
+                "delete_on_termination": "bool",
+                "encrypted": "bool",
+                "iops": "number",
+                "kms_key_id": "string",
+                "volume_id": "string",
+                "volume_size": "number",
+                "volume_type": "string"
+              }
+            ]
+          ],
+          "security_groups": [
+            "set",
+            "string"
+          ],
+          "source_dest_check": "bool",
+          "subnet_id": "string",
+          "tags": [
+            "map",
+            "string"
+          ],
+          "tenancy": "string",
+          "timeouts": [
+            "object",
+            {
+              "create": "string",
+              "delete": "string",
+              "update": "string"
+            }
+          ],
+          "user_data": "string",
+          "user_data_base64": "string",
+          "volume_tags": [
+            "map",
+            "string"
+          ],
+          "vpc_security_group_ids": [
+            "set",
+            "string"
+          ]
+        }
+      ]
+    },
+    "public_ip": {
+      "value": "18.185.127.109",
+      "type": "string"
+    }
+  },
+  "resources": [
+    {
+      "mode": "managed",
+      "type": "aws_instance",
+      "name": "shreya_webserver",
+      "provider": "provider.aws",
+      "instances": [
+        {
+          "schema_version": 1,
+          "attributes": {
+            "ami": "ami-00aa4671cbf840d82",
+            "arn": "arn:aws:ec2:eu-central-1:426714351745:instance/i-02819c8496758a607",
+            "associate_public_ip_address": true,
+            "availability_zone": "eu-central-1b",
+            "cpu_core_count": 1,
+            "cpu_threads_per_core": 1,
+            "credit_specification": [
+              {
+                "cpu_credits": "standard"
+              }
+            ],
+            "disable_api_termination": false,
+            "ebs_block_device": [],
+            "ebs_optimized": false,
+            "ephemeral_block_device": [],
+            "get_password_data": false,
+            "host_id": null,
+            "iam_instance_profile": "",
+            "id": "i-02819c8496758a607",
+            "instance_initiated_shutdown_behavior": null,
+            "instance_state": "running",
+            "instance_type": "t2.micro",
+            "ipv6_address_count": 0,
+            "ipv6_addresses": [],
+            "key_name": "",
+            "monitoring": false,
+            "network_interface": [],
+            "network_interface_id": null,
+            "password_data": "",
+            "placement_group": "",
+            "primary_network_interface_id": "eni-049aeb7aea7b8e2af",
+            "private_dns": "ip-172-31-42-116.eu-central-1.compute.internal",
+            "private_ip": "172.31.42.116",
+            "public_dns": "ec2-18-185-127-109.eu-central-1.compute.amazonaws.com",
+            "public_ip": "18.185.127.109",
+            "root_block_device": [
+              {
+                "delete_on_termination": true,
+                "encrypted": false,
+                "iops": 100,
+                "kms_key_id": "",
+                "volume_id": "vol-072498c51d755da1c",
+                "volume_size": 8,
+                "volume_type": "gp2"
+              }
+            ],
+            "security_groups": [
+              "default"
+            ],
+            "source_dest_check": true,
+            "subnet_id": "subnet-03e6e87e",
+            "tags": {
+              "Name": "ShreyaSpace",
+              "owner": "shreya"
+            },
+            "tenancy": "default",
+            "timeouts": null,
+            "user_data": null,
+            "user_data_base64": null,
+            "volume_tags": {},
+            "vpc_security_group_ids": [
+              "sg-395d185a"
+            ]
+          },
+          "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6MTIwMDAwMDAwMDAwMCwidXBkYXRlIjo2MDAwMDAwMDAwMDB9LCJzY2hlbWFfdmVyc2lvbiI6IjEifQ=="
+        }
+      ]
+    },
+    {
+      "mode": "managed",
+      "type": "aws_s3_bucket",
+      "name": "shreyatfstate",
+      "provider": "provider.aws",
+      "instances": [
+        {
+          "schema_version": 0,
+          "attributes": {
+            "acceleration_status": "",
+            "acl": "private",
+            "arn": "arn:aws:s3:::shreyaspace-tfstore",
+            "bucket": "shreyaspace-tfstore",
+            "bucket_domain_name": "shreyaspace-tfstore.s3.amazonaws.com",
+            "bucket_prefix": null,
+            "bucket_regional_domain_name": "shreyaspace-tfstore.s3.eu-central-1.amazonaws.com",
+            "cors_rule": [],
+            "force_destroy": false,
+            "hosted_zone_id": "Z21DNDUVLTQW6Q",
+            "id": "shreyaspace-tfstore",
+            "lifecycle_rule": [],
+            "logging": [],
+            "object_lock_configuration": [],
+            "policy": null,
+            "region": "eu-central-1",
+            "replication_configuration": [],
+            "request_payer": "BucketOwner",
+            "server_side_encryption_configuration": [],
+            "tags": {
+              "Name": "Shreya bucket",
+              "owner": "shreya"
+            },
+            "versioning": [
+              {
+                "enabled": false,
+                "mfa_delete": false
+              }
+            ],
+            "website": [],
+            "website_domain": null,
+            "website_endpoint": null
+          },
+          "private": "bnVsbA=="
+        }
+      ]
+    }
+  ]
+}
diff --git a/terraform/shreyach13/terraform.tfstate.backup b/terraform/shreyach13/terraform.tfstate.backup
new file mode 100644
index 0000000..063a6e9
--- /dev/null
+++ b/terraform/shreyach13/terraform.tfstate.backup
@@ -0,0 +1,331 @@
+{
+  "version": 4,
+  "terraform_version": "0.12.9",
+  "serial": 16,
+  "lineage": "37933da2-09f3-c391-ce85-b57d0901dacc",
+  "outputs": {
+    "public_ip": {
+      "value": "18.185.127.109",
+      "type": "string"
+    },
+    "shreya_webserver": {
+      "value": {
+        "ami": "ami-00aa4671cbf840d82",
+        "arn": "arn:aws:ec2:eu-central-1:426714351745:instance/i-02819c8496758a607",
+        "associate_public_ip_address": true,
+        "availability_zone": "eu-central-1b",
+        "cpu_core_count": 1,
+        "cpu_threads_per_core": 1,
+        "credit_specification": [
+          {
+            "cpu_credits": "standard"
+          }
+        ],
+        "disable_api_termination": false,
+        "ebs_block_device": [],
+        "ebs_optimized": false,
+        "ephemeral_block_device": [],
+        "get_password_data": false,
+        "host_id": null,
+        "iam_instance_profile": "",
+        "id": "i-02819c8496758a607",
+        "instance_initiated_shutdown_behavior": null,
+        "instance_state": "running",
+        "instance_type": "t2.micro",
+        "ipv6_address_count": 0,
+        "ipv6_addresses": [],
+        "key_name": "",
+        "monitoring": false,
+        "network_interface": [],
+        "network_interface_id": null,
+        "password_data": "",
+        "placement_group": "",
+        "primary_network_interface_id": "eni-049aeb7aea7b8e2af",
+        "private_dns": "ip-172-31-42-116.eu-central-1.compute.internal",
+        "private_ip": "172.31.42.116",
+        "public_dns": "ec2-18-185-127-109.eu-central-1.compute.amazonaws.com",
+        "public_ip": "18.185.127.109",
+        "root_block_device": [
+          {
+            "delete_on_termination": true,
+            "encrypted": false,
+            "iops": 100,
+            "kms_key_id": "",
+            "volume_id": "vol-072498c51d755da1c",
+            "volume_size": 8,
+            "volume_type": "gp2"
+          }
+        ],
+        "security_groups": [
+          "default"
+        ],
+        "source_dest_check": true,
+        "subnet_id": "subnet-03e6e87e",
+        "tags": {
+          "Name": "ShreyaSpace",
+          "owner": "shreya"
+        },
+        "tenancy": "default",
+        "timeouts": null,
+        "user_data": null,
+        "user_data_base64": null,
+        "volume_tags": {},
+        "vpc_security_group_ids": [
+          "sg-395d185a"
+        ]
+      },
+      "type": [
+        "object",
+        {
+          "ami": "string",
+          "arn": "string",
+          "associate_public_ip_address": "bool",
+          "availability_zone": "string",
+          "cpu_core_count": "number",
+          "cpu_threads_per_core": "number",
+          "credit_specification": [
+            "list",
+            [
+              "object",
+              {
+                "cpu_credits": "string"
+              }
+            ]
+          ],
+          "disable_api_termination": "bool",
+          "ebs_block_device": [
+            "set",
+            [
+              "object",
+              {
+                "delete_on_termination": "bool",
+                "device_name": "string",
+                "encrypted": "bool",
+                "iops": "number",
+                "kms_key_id": "string",
+                "snapshot_id": "string",
+                "volume_id": "string",
+                "volume_size": "number",
+                "volume_type": "string"
+              }
+            ]
+          ],
+          "ebs_optimized": "bool",
+          "ephemeral_block_device": [
+            "set",
+            [
+              "object",
+              {
+                "device_name": "string",
+                "no_device": "bool",
+                "virtual_name": "string"
+              }
+            ]
+          ],
+          "get_password_data": "bool",
+          "host_id": "string",
+          "iam_instance_profile": "string",
+          "id": "string",
+          "instance_initiated_shutdown_behavior": "string",
+          "instance_state": "string",
+          "instance_type": "string",
+          "ipv6_address_count": "number",
+          "ipv6_addresses": [
+            "list",
+            "string"
+          ],
+          "key_name": "string",
+          "monitoring": "bool",
+          "network_interface": [
+            "set",
+            [
+              "object",
+              {
+                "delete_on_termination": "bool",
+                "device_index": "number",
+                "network_interface_id": "string"
+              }
+            ]
+          ],
+          "network_interface_id": "string",
+          "password_data": "string",
+          "placement_group": "string",
+          "primary_network_interface_id": "string",
+          "private_dns": "string",
+          "private_ip": "string",
+          "public_dns": "string",
+          "public_ip": "string",
+          "root_block_device": [
+            "list",
+            [
+              "object",
+              {
+                "delete_on_termination": "bool",
+                "encrypted": "bool",
+                "iops": "number",
+                "kms_key_id": "string",
+                "volume_id": "string",
+                "volume_size": "number",
+                "volume_type": "string"
+              }
+            ]
+          ],
+          "security_groups": [
+            "set",
+            "string"
+          ],
+          "source_dest_check": "bool",
+          "subnet_id": "string",
+          "tags": [
+            "map",
+            "string"
+          ],
+          "tenancy": "string",
+          "timeouts": [
+            "object",
+            {
+              "create": "string",
+              "delete": "string",
+              "update": "string"
+            }
+          ],
+          "user_data": "string",
+          "user_data_base64": "string",
+          "volume_tags": [
+            "map",
+            "string"
+          ],
+          "vpc_security_group_ids": [
+            "set",
+            "string"
+          ]
+        }
+      ]
+    }
+  },
+  "resources": [
+    {
+      "mode": "managed",
+      "type": "aws_instance",
+      "name": "shreya_webserver",
+      "provider": "provider.aws",
+      "instances": [
+        {
+          "schema_version": 1,
+          "attributes": {
+            "ami": "ami-00aa4671cbf840d82",
+            "arn": "arn:aws:ec2:eu-central-1:426714351745:instance/i-02819c8496758a607",
+            "associate_public_ip_address": true,
+            "availability_zone": "eu-central-1b",
+            "cpu_core_count": 1,
+            "cpu_threads_per_core": 1,
+            "credit_specification": [
+              {
+                "cpu_credits": "standard"
+              }
+            ],
+            "disable_api_termination": false,
+            "ebs_block_device": [],
+            "ebs_optimized": false,
+            "ephemeral_block_device": [],
+            "get_password_data": false,
+            "host_id": null,
+            "iam_instance_profile": "",
+            "id": "i-02819c8496758a607",
+            "instance_initiated_shutdown_behavior": null,
+            "instance_state": "running",
+            "instance_type": "t2.micro",
+            "ipv6_address_count": 0,
+            "ipv6_addresses": [],
+            "key_name": "",
+            "monitoring": false,
+            "network_interface": [],
+            "network_interface_id": null,
+            "password_data": "",
+            "placement_group": "",
+            "primary_network_interface_id": "eni-049aeb7aea7b8e2af",
+            "private_dns": "ip-172-31-42-116.eu-central-1.compute.internal",
+            "private_ip": "172.31.42.116",
+            "public_dns": "ec2-18-185-127-109.eu-central-1.compute.amazonaws.com",
+            "public_ip": "18.185.127.109",
+            "root_block_device": [
+              {
+                "delete_on_termination": true,
+                "encrypted": false,
+                "iops": 100,
+                "kms_key_id": "",
+                "volume_id": "vol-072498c51d755da1c",
+                "volume_size": 8,
+                "volume_type": "gp2"
+              }
+            ],
+            "security_groups": [
+              "default"
+            ],
+            "source_dest_check": true,
+            "subnet_id": "subnet-03e6e87e",
+            "tags": {
+              "Name": "ShreyaSpace",
+              "owner": "shreya"
+            },
+            "tenancy": "default",
+            "timeouts": null,
+            "user_data": null,
+            "user_data_base64": null,
+            "volume_tags": {},
+            "vpc_security_group_ids": [
+              "sg-395d185a"
+            ]
+          },
+          "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6MTIwMDAwMDAwMDAwMCwidXBkYXRlIjo2MDAwMDAwMDAwMDB9LCJzY2hlbWFfdmVyc2lvbiI6IjEifQ=="
+        }
+      ]
+    },
+    {
+      "mode": "managed",
+      "type": "aws_s3_bucket",
+      "name": "shreyatfstate",
+      "provider": "provider.aws",
+      "instances": [
+        {
+          "schema_version": 0,
+          "attributes": {
+            "acceleration_status": "",
+            "acl": "private",
+            "arn": "arn:aws:s3:::shreyaspace-tfstore",
+            "bucket": "shreyaspace-tfstore",
+            "bucket_domain_name": "shreyaspace-tfstore.s3.amazonaws.com",
+            "bucket_prefix": null,
+            "bucket_regional_domain_name": "shreyaspace-tfstore.s3.eu-central-1.amazonaws.com",
+            "cors_rule": [],
+            "force_destroy": false,
+            "hosted_zone_id": "Z21DNDUVLTQW6Q",
+            "id": "shreyaspace-tfstore",
+            "lifecycle_rule": [],
+            "logging": [],
+            "object_lock_configuration": [],
+            "policy": null,
+            "region": "eu-central-1",
+            "replication_configuration": [],
+            "request_payer": "BucketOwner",
+            "server_side_encryption_configuration": [],
+            "tags": {
+              "Name": "Shreya bucket",
+              "owner": "shreya"
+            },
+            "versioning": [
+              {
+                "enabled": false,
+                "mfa_delete": false
+              }
+            ],
+            "website": [],
+            "website_domain": null,
+            "website_endpoint": null
+          },
+          "private": "bnVsbA=="
+        }
+      ]
+    }
+  ]
+}
diff --git a/terraform/shreyach13/variable.tf b/terraform/shreyach13/variable.tf
new file mode 100644
index 0000000..df4cbc6
--- /dev/null
+++ b/terraform/shreyach13/variable.tf
@@ -0,0 +1,3 @@
+variable "owner" {
+  default = "shreya"
+}
diff --git a/terraform/shreyach13/with-remote-state/app/backend.hcl b/terraform/shreyach13/with-remote-state/app/backend.hcl
new file mode 100644
index 0000000..9896b80
--- /dev/null
+++ b/terraform/shreyach13/with-remote-state/app/backend.hcl
@@ -0,0 +1,3 @@
+bucket = "shreyaspace-tfstore"
+region = "eu-central-1"
+encrypt = true
diff --git a/terraform/shreyach13/with-remote-state/app/backend.tf b/terraform/shreyach13/with-remote-state/app/backend.tf
new file mode 100644
index 0000000..cdba228
--- /dev/null
+++ b/terraform/shreyach13/with-remote-state/app/backend.tf
@@ -0,0 +1,5 @@
+terraform {
+  backend "s3" {
+    key = "core/terraform.tfstate"
+  }
+}
diff --git a/terraform/shreyach13/with-remote-state/app/main.tf b/terraform/shreyach13/with-remote-state/app/main.tf
new file mode 100644
index 0000000..62939c2
--- /dev/null
+++ b/terraform/shreyach13/with-remote-state/app/main.tf
@@ -0,0 +1,14 @@
+provider "aws" {
+  region = "eu-central-1"
+}
+
+module "shreya_web_server" {
+  source = "../modules/compute"
+  server_port = var.server_port
+}
+
+module "security" {
+  source = "../modules/security"
+  from_port = var.server_port
+  to_port = var.server_port
+}
diff --git a/terraform/shreyach13/with-remote-state/app/tf.plan b/terraform/shreyach13/with-remote-state/app/tf.plan
new file mode 100644
index 0000000..c7dd358
Binary files /dev/null and b/terraform/shreyach13/with-remote-state/app/tf.plan differ
diff --git a/terraform/shreyach13/with-remote-state/app/variable.tf b/terraform/shreyach13/with-remote-state/app/variable.tf
new file mode 100644
index 0000000..73bf34f
--- /dev/null
+++ b/terraform/shreyach13/with-remote-state/app/variable.tf
@@ -0,0 +1,4 @@
+variable "server_port" {
+  description = "This port will be used for http requests"
+  default = "8080"
+}
diff --git a/terraform/shreyach13/with-remote-state/modules/compute/main.tf b/terraform/shreyach13/with-remote-state/modules/compute/main.tf
new file mode 100644
index 0000000..c347c0a
--- /dev/null
+++ b/terraform/shreyach13/with-remote-state/modules/compute/main.tf
@@ -0,0 +1,18 @@
+provider "aws" {
+  region = "eu-central-1"
+}
+
+# A basic vm example to show how to provision a vm
+resource "aws_instance" "shreya_webserver"{
+  ami = "ami-00aa4671cbf840d82"
+  instance_type = "t2.micro"
+  user_data = <<-EOF
+            #!/bin/bash
+            echo "Hello, World" >index.html
+            nohup busybox httpd -f -p "${var.server_port}" &
+            EOF
+  #This will change the name of the instance
+  tags = {
+    Name = "ShreyaSpace"
+  }
+}
diff --git a/terraform/shreyach13/with-remote-state/modules/compute/output.tf b/terraform/shreyach13/with-remote-state/modules/compute/output.tf
new file mode 100644
index 0000000..f638f79
--- /dev/null
+++ b/terraform/shreyach13/with-remote-state/modules/compute/output.tf
@@ -0,0 +1,4 @@
+output "public_ip" {
+  description="the public ip of the server"
+  value=aws_instance.shreya_webserver.public_ip
+}
diff --git a/terraform/shreyach13/with-remote-state/modules/compute/variable.tf b/terraform/shreyach13/with-remote-state/modules/compute/variable.tf
new file mode 100644
index 0000000..73bf34f
--- /dev/null
+++ b/terraform/shreyach13/with-remote-state/modules/compute/variable.tf
@@ -0,0 +1,4 @@
+variable "server_port" {
+  description = "This port will be used for http requests"
+  default = "8080"
+}
diff --git a/terraform/shreyach13/with-remote-state/modules/security/main.tf b/terraform/shreyach13/with-remote-state/modules/security/main.tf
new file mode 100644
index 0000000..51e2696
--- /dev/null
+++ b/terraform/shreyach13/with-remote-state/modules/security/main.tf
@@ -0,0 +1,13 @@
+resource "aws_security_group" "shsecurity"{
+  name = "terraform-instance-${random_id.myrandom.hex}"
+  ingress {
+    from_port = var.from_port
+    to_port = var.to_port
+    protocol = "tcp"
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+}
+
+resource "random_id" "myrandom" {
+  byte_length = 3
+}
diff --git a/terraform/shreyach13/with-remote-state/modules/security/variable.tf b/terraform/shreyach13/with-remote-state/modules/security/variable.tf
new file mode 100644
index 0000000..398d418
--- /dev/null
+++ b/terraform/shreyach13/with-remote-state/modules/security/variable.tf
@@ -0,0 +1,7 @@
+variable "from_port" {
+  default = "9000"
+}
+
+variable "to_port" {
+  default = "9000"
+}
